### ISSUE TITLE
Variable is defined still I am getting TS2454 Error

### ISSUE THREAD

**Issue by tbhaxor** [microsoft/TypeScript#42143]
# Bug Report

### 🔎 Search Terms

1. TS2454 false positive
2. TS2454 error even if variable is already defined

### 🕗 Version & Regression Information

False Positive

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCLAM4IPYFdzIDeAUMmcsACYBcyGYUoA5qefYwA4QCSNdDzYgF9iomPgRhg2EMiYQwAYTwcZACiq12zAJS10WPAQA+yELgA2FoqzJQFuKLPNXhogG5woyGBbhNkAF4fOAsMCFELBWQcXFUQfUxYkzNLC2JgGGQ1X38dG3IYlRkguQVlOPUAIgBlAHkAWQBRZG5UKp03SOiEUtj4oA)

### 💻 Code

```ts
let coupon: Discount | null // ALREADY ASSIGNED HERE
if (req.body.coupon) {
	coupon = await DISCOUNT.findOne({
		isExpired: false,
		_id: req.body.coupon
	})
	if (!coupon) {
		return res.json({
			success: false,
			message: 'Coupon not found'
		})
	}
	invoice.discount = coupon.id
	price = applyDiscount(price, coupon.type, coupon.amount)
}
invoice.amount = price

const ss = await stripe.subscriptions.create({
	customer: user.stripeId,
	coupon: coupon ? .stripeId,
	collection_method: 'charge_automatically',
	items: [{
		plan: plan.stripeId
	}],
	metadata: {
		subscription: subscription.id as string
	},
	payment_behavior: 'error_if_incomplete',
})

subscription.stripeId = ss.id

await subscription.save()
```

### 🙁 Actual behavior

![image](https://user-images.githubusercontent.com/28386721/103262033-afb41180-49c9-11eb-9b8f-31dac43fa45c.png)

### 🙂 Expected behavior

It should not be coming :laughing:

**Comment by MartinJohns** [comment]
> let coupon: Discount | null // ALREADY ASSIGNED HERE

You have it declared, but you're not assigning something in all code paths.

**Comment by tbhaxor** [comment]
Ohk thanks @MartinJohns 

Also for those who are new in JS and TS

By default the variable will be `undefined` during the declaration, and null itself is a value

So according to this
```ts
let coupon: Discount | null // error
let coupon: Discount | null = null // fixed

let coupon: Discount | undefined // no error
```

**Comment by MartinJohns** [comment]
> By default the variable will be `undefined` during the declaration, and `null` itself is a value

Yep. But be aware that in TypeScript `undefined` and `null` only exist as types when you have the compiler option `strictNullChecks` enabled (which unfortunately many frameworks still don't have).
